### ISSUE TITLE
Increasing speed of polyA tail length determination

### ISSUE THREAD

**Issue by maximus-sci** [jts/nanopolish#967]
Hello,

I'm currently trying to run nanopolish polya and it's quite slow (taking nearly a day for a file with ~1 million reads). First, I'm wondering if this is normal and, if so, what the bottleneck likely is. I am running this on a computing cluster, so I can increase the memory and number of threads to whatever it needs to be.

I'm also wondering how the qc_tag field interacts with the MinKNOW software's adaptive sampling capabilities. Presumably a few hundred bases of a nucleic acid will already have been read by the sequencer before it's able to reject the read and spit it back out. I'm guessing this would result in a similar signature as when there's a mux change to unclog a blocked pore. Would the QC tag for these reads still be a "PASS" or something else?

Since I'm just measuring polyA tails, the 5' end of the read isn't that important for me.

**Comment by jts** [comment]
Hi @maximus-sci,

> I'm currently trying to run nanopolish polya and it's quite slow (taking nearly a day for a file with ~1 million reads). First, I'm wondering if this is normal and, if so, what the bottleneck likely is. I am running this on a computing cluster, so I can increase the memory and number of threads to whatever it needs to be.

Since you a computing cluster, I suggest splitting up the input into chunks (say 50,000 reads) and process each in parallel with a separate job. This is better than increasing the number of threads within a single job, as nanopolish is limited by the HDF5. 

> I'm also wondering how the qc_tag field interacts with the MinKNOW software's adaptive sampling capabilities. Presumably a few hundred bases of a nucleic acid will already have been read by the sequencer before it's able to reject the read and spit it back out. I'm guessing this would result in a similar signature as when there's a mux change to unclog a blocked pore. Would the QC tag for these reads still be a "PASS" or something else?

I don't know how this will interact with adaptive sampling, sorry. It will depend on exactly how the rejection decision is made, and how much of the molecule is sequenced.

**Comment by maximus-sci** [comment]
Would I have to split it up from the beginning (like at the basecalling and alignment stage)? 

That's the only way that I can think to keep everything straight

**Comment by hasindu2008** [comment]
@jts  
Sorry for hacking this threads :) Saw this question accidentally. 

@maximus-sci 
Given that the nanopolish master branch now has SLOW5 support, converting all the FAST5 to SLOW5 first using [slow5tools](https://github.com/hasindu2008/slow5tools) and then running nanopolish on SLOW5 might help alleviate this bottleneck.  The overhead for conversion for a 20 million set of reads on our cluster file system with 48 cores was 5h. Check the nanopolish example under https://hasindu2008.github.io/slow5tools/workflows.html. That way, there won't be a need to manually split the data. You could give it a try.

**Comment by jts** [comment]
That is a good suggestion, thanks Hasindu!

**Comment by maximus-sci** [comment]
This is a very cool tool, thank you!
Is there any information lost upon conversion to SLOW5/BLOW5? I am wondering if, after conversion, I still need to hang on to the FAST5 files?

I don't fully understand the issues, but I guess the library required to read the hdf5 format is limited in some way that creates a computational bottleneck?

**Comment by hasindu2008** [comment]
@maximus-sci 

SLOW5/BLOW5 retains all the information from FAST5 and you can even convert SLOW5/BLOW5 files back to FAST5 using slow5tools slow52fast5 subtool.
Yeh, it is an issue with the HDF5 library that serialises the threads going through it.  SLOW5 preprint has more information about this https://www.biorxiv.org/content/10.1101/2021.06.29.450255v1 and a more detailed technical description is in the supplementary note https://www.biorxiv.org/content/biorxiv/early/2021/06/30/2021.06.29.450255/DC2/embed/media-2.pdf?download=true

**Comment by maximus-sci** [comment]
Hello @hasindu2008 I'm unable to use slow5 files for polyA calling and was hoping you could help.

I can successfully convert my fast5 files into slow5 and then merge them together using slow5tools.
Indexing with nanopolish also executes with no errors. 

When I try to call polyA tail lengths, the script runs with no errors and produces this output:
`$nanopolish polya --reads .merged.fastq.gz --bam merged.sorted.bam --genome Reference.cdna.all.fa.bgzf > polyA.tsv
[post-run summary] total reads: 28975, unparseable: 0, qc fail: 365, could not calibrate: 524, no alignment: 56, bad fast5: 0`

but in the tsv file, every single read has the qc tag "READ_FAILED_LOAD"

When I try to get tail lengths without converting to slow5, many/most of the reads have the qc tag PASS.
Any advice?

**Comment by hasindu2008** [comment]
@maximus-sci 

It is something in the polya subprogramme that we overlooked when integrating SLOW5. I have applied a fix in my fork at https://github.com/hasindu2008/nanopolish-arm/tree/slow5.

You can try as:
````
git clone --recursive https://github.com/hasindu2008/nanopolish-arm -b slow5
make
````

@jts 
Seems like the polya is using `sr.fast5_path == ""` to see if signal data was loaded correctly. So as a workaround I have set the fast5_path to "slow5" in the SLOW5 mode.
If you are happy with the fix and if @maximus-sci gets it working, I will send a pull request. Let me know if you have any other suggestions.
https://github.com/jts/nanopolish/compare/master...hasindu2008:slow5

**Comment by maximus-sci** [comment]
@hasindu2008 

the nanopolish-arm branch that you made does run! 

In my test run, calling tail lengths using the slow5 files took about 1/2 the time compared to the standard way with fast5 files. Thanks so much.

**Comment by hasindu2008** [comment]
@maximus-sci 
Let me know the performance for your large dataset on the cluster goes - on our cluster for eventalign and methcalling there was like a 5X improvement. I have not tested for polyA so I am interested to know.
